File: tests/unit/license.test.ts
import { afterAll, beforeAll, describe, expect, it } from 'vitest';
import { detectSpdx, findLicenseEntry, isLicenseFilename, resolveLicense } from '../../src/lib/ingest/license';
import { findReadmeEntry } from '../../src/lib/ingest/readme';
import { scanRepo } from '../../src/lib/ingest/scan';
import { listRootTree } from '../../src/lib/ingest/tree';
import { FixtureRepo } from './helpers/fixture-repo';

const MIT = `MIT License

Copyright (c) 2024 Someone

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction...
`;

const APACHE = `                                 Apache License
                           Version 2.0, January 2004
                        http://www.apache.org/licenses/

   TERMS AND CONDITIONS FOR USE, REPRODUCTION, AND DISTRIBUTION
`;

const GPL3 = `                    GNU GENERAL PUBLIC LICENSE
                       Version 3, 29 June 2007
`;

const BSD3 = `Redistribution and use in source and binary forms, with or without
modification, are permitted provided that the following conditions are met:
3. Neither the name of the copyright holder nor the names of its
   contributors may be used to endorse or promote products derived from
   this software without specific prior written permission.
`;

const opts = { maxBlobBytes: 1024 * 1024, maxCommits: null };

describe('license detection', () => {
  it('recognises common license texts', () => {
    expect(detectSpdx(MIT)).toBe('MIT');
    expect(detectSpdx(APACHE)).toBe('Apache-2.0');
    expect(detectSpdx(GPL3)).toBe('GPL-3.0-only');
    expect(detectSpdx('GNU LESSER GENERAL PUBLIC LICENSE\nVersion 3, 29 June 2007')).toBe('LGPL-3.0-only');
    expect(detectSpdx('GNU AFFERO GENERAL PUBLIC LICENSE\nVersion 3, 19 November 2007')).toBe('AGPL-3.0-only');
    expect(detectSpdx('GNU GENERAL PUBLIC LICENSE\nVersion 2, June 1991')).toBe('GPL-2.0-only');
    expect(detectSpdx(BSD3)).toBe('BSD-3-Clause');
    expect(detectSpdx('Redistribution and use in source and binary forms, with or without modification, are permitted')).toBe('BSD-2-Clause');
    expect(detectSpdx('Mozilla Public License Version 2.0\n==================================')).toBe('MPL-2.0');
    expect(detectSpdx('ISC License\n\nCopyright (c) 2024')).toBe('ISC');
    expect(detectSpdx('This is free and unencumbered software released into the public domain.')).toBe('Unlicense');
    expect(detectSpdx('CC0 1.0 Universal')).toBe('CC0-1.0');
    expect(detectSpdx('Zero-Clause BSD\n=============')).toBe('0BSD');
    expect(detectSpdx('All rights reserved. Do not copy.')).toBeNull();
  });

  it('matches license filenames case-insensitively', () => {
    for (const n of ['LICENSE', 'license', 'LICENSE.md', 'LICENSE.txt', 'LICENCE', 'COPYING', 'COPYING.LESSER', 'UNLICENSE', 'LICENSE-MIT']) {
      expect(isLicenseFilename(n), n).toBe(true);
    }
    expect(isLicenseFilename('LICENSES.md')).toBe(false);
    expect(isLicenseFilename('README.md')).toBe(false);
  });

  it('resolveLicense honours config override', () => {
    expect(resolveLicense({ file: 'LICENSE', spdx: 'MIT' }, null)).toEqual({ spdx: 'MIT', file: 'LICENSE', source: 'file' });
    expect(resolveLicense({ file: 'LICENSE', spdx: null }, 'Apache-2.0')).toEqual({ spdx: 'Apache-2.0', file: 'LICENSE', source: 'config' });
    expect(resolveLicense(null, 'MIT')).toEqual({ spdx: 'MIT', file: null, source: 'config' });
    expect(resolveLicense(null, null)).toBeNull();
  });
});

describe('license + readme in a repo', () => {
  let mit: FixtureRepo;
  let apache: FixtureRepo;
  let none: FixtureRepo;

  beforeAll(() => {
    mit = FixtureRepo.create('mit');
    mit.writeAndCommit({ LICENSE: MIT, 'README.md': '# MIT repo\n', README: 'plain readme\n', 'readme.txt': 'txt\n' }, 'init');
    apache = FixtureRepo.create('apache');
    apache.writeAndCommit({ 'LICENSE.txt': APACHE, 'readme.md': '# lower\n' }, 'init');
    none = FixtureRepo.create('none');
    none.writeAndCommit({ 'src/LICENSE': MIT, 'docs/README.md': '# nested\n' }, 'init');
  });
  afterAll(() => {
    mit.cleanup();
    apache.cleanup();
    none.cleanup();
  });

  it('finds root license + readme entries, preferring README.md', async () => {
    const root = await listRootTree(mit.dir, 'main');
    expect(findLicenseEntry(root)!.path).toBe('LICENSE');
    expect(findReadmeEntry(root)!.path).toBe('README.md');
    const apRoot = await listRootTree(apache.dir, 'main');
    expect(findLicenseEntry(apRoot)!.path).toBe('LICENSE.txt');
    expect(findReadmeEntry(apRoot)!.path).toBe('readme.md');
    const noneRoot = await listRootTree(none.dir, 'main');
    expect(findLicenseEntry(noneRoot)).toBeNull();
    expect(findReadmeEntry(noneRoot)).toBeNull();
  });

  it('scanRepo reports detected licenses and readme content', async () => {
    const a = await scanRepo({ absPath: mit.dir }, opts);
    if ('skipped' in a) throw new Error('skipped');
    expect(a.repo.license).toEqual({ spdx: 'MIT', file: 'LICENSE', source: 'file' });
    expect(a.repo.readme).toMatchObject({ path: 'README.md', content: '# MIT repo\n' });
    expect(a.repo.readme!.sha).toBe(a.repo.files['README.md']!.sha);

    const b = await scanRepo({ absPath: apache.dir }, opts);
    if ('skipped' in b) throw new Error('skipped');
    expect(b.repo.license).toEqual({ spdx: 'Apache-2.0', file: 'LICENSE.txt', source: 'file' });
    expect(b.repo.readme!.path).toBe('readme.md');

    const c = await scanRepo({ absPath: none.dir }, opts);
    if ('skipped' in c) throw new Error('skipped');
    expect(c.repo.license).toBeNull();
    expect(c.repo.readme).toBeNull();
  });

  it('config override wins but keeps the detected file', async () => {
    const a = await scanRepo({ absPath: mit.dir, overrides: { license: 'BSD-3-Clause' } }, opts);
    if ('skipped' in a) throw new Error('skipped');
    expect(a.repo.license).toEqual({ spdx: 'BSD-3-Clause', file: 'LICENSE', source: 'config' });
    const c = await scanRepo({ absPath: none.dir, overrides: { license: 'MIT' } }, opts);
    if ('skipped' in c) throw new Error('skipped');
    expect(c.repo.license).toEqual({ spdx: 'MIT', file: null, source: 'config' });
  });

  it('readme over the blob size cap is null', async () => {
    const a = await scanRepo({ absPath: mit.dir }, { maxBlobBytes: 4, maxCommits: null });
    if ('skipped' in a) throw new Error('skipped');
    expect(a.repo.readme).toBeNull();
    expect(a.repo.license).toEqual({ spdx: 'MIT', file: 'LICENSE', source: 'file' });
  });
});

File: internal/ingest/license_test.go
package ingest

import (
	"strings"
	"testing"
)

func TestIsLicenseFilename(t *testing.T) {
	for name, want := range map[string]bool{
		"LICENSE":        true,
		"license":        true,
		"LICENSE.md":     true,
		"LICENSE.txt":    true,
		"LICENSE-MIT":    true,
		"licence":        true,
		"LICENCE.md":     true,
		"licence-apache": true,
		"COPYING":        true,
		"copying.txt":    true,
		"UNLICENSE":      true,
		"unlicense.txt":  true,
		// A prefix match needs the separator: only "license." and "license-" continue the name.
		"licenses":      false,
		"license_old":   false,
		"copying-extra": false,
		"unlicense.md":  false,
		"mylicense":     false,
		"readme.md":     false,
		"":              false,
	} {
		if got := isLicenseFilename(name); got != want {
			t.Errorf("isLicenseFilename(%q) = %v, want %v", name, got, want)
		}
	}
}

func TestFindLicenseEntry(t *testing.T) {
	blob := func(name string) RawTreeEntry {
		return RawTreeEntry{Type: "blob", Name: name, Path: name, Sha: "sha-" + name}
	}
	cases := []struct {
		name    string
		entries []RawTreeEntry
		want    string // "" means nil
	}{
		{"none", []RawTreeEntry{blob("README.md"), blob("main.go")}, ""},
		// A plain LICENSE outranks every qualified spelling, whatever order git listed them in.
		{"plain wins", []RawTreeEntry{blob("LICENSE-MIT"), blob("COPYING"), blob("LICENSE")}, "LICENSE"},
		{"license.md is still plain", []RawTreeEntry{blob("LICENSE-MIT"), blob("LICENSE.md")}, "LICENSE.md"},
		// Same rank: the name breaks the tie, in code-point order.
		{"rank tie", []RawTreeEntry{blob("LICENSE-MIT"), blob("LICENSE-APACHE")}, "LICENSE-APACHE"},
		{"licence beats copying", []RawTreeEntry{blob("COPYING"), blob("LICENCE.md")}, "LICENCE.md"},
		// A directory named LICENSE is not a license file.
		{"blobs only", []RawTreeEntry{{Type: "tree", Name: "LICENSE", Path: "LICENSE"}}, ""},
	}
	for _, c := range cases {
		got := findLicenseEntry(c.entries)
		switch {
		case c.want == "" && got != nil:
			t.Errorf("%s: got %q, want none", c.name, got.Name)
		case c.want != "" && got == nil:
			t.Errorf("%s: got none, want %q", c.name, c.want)
		case c.want != "" && got.Name != c.want:
			t.Errorf("%s: got %q, want %q", c.name, got.Name, c.want)
		}
	}
}

func TestDetectSpdx(t *testing.T) {
	for text, want := range map[string]string{
		mitLicense: "MIT",
		"Permission is hereby granted, free of charge, to any person obtaining a copy\nof this software\n": "MIT",
		"ISC License\n\nCopyright (c) 2024\n":                                                                "ISC",
		"                Apache License\n         Version 2.0, January 2004\n":                               "Apache-2.0",
		"Mozilla Public License Version 2.0\n":                                                               "MPL-2.0",
		"GNU AFFERO GENERAL PUBLIC LICENSE\nVersion 3, 19 November 2007\n":                                   "AGPL-3.0-only",
		"GNU LESSER GENERAL PUBLIC LICENSE\nVersion 3, 29 June 2007\n":                                       "LGPL-3.0-only",
		"GNU LESSER GENERAL PUBLIC LICENSE\nVersion 2.1, February 1999\n":                                    "LGPL-2.1-only",
		"GNU GENERAL PUBLIC LICENSE\nVersion 3, 29 June 2007\n":                                              "GPL-3.0-only",
		"GNU GENERAL PUBLIC LICENSE\nVersion 2, June 1991\n":                                                 "GPL-2.0-only",
		"This is free and unencumbered software released into the public domain.\n":                          "Unlicense",
		"Creative Commons Legal Code\n\nCC0 1.0 Universal\n":                                                 "CC0-1.0",
		"Redistribution and use in source and binary forms, with or without\nmodification, are permitted.\n": "BSD-2-Clause",
		"Redistribution and use in source and binary forms are permitted provided that\n" +
			"neither the name of the holder nor the names of its contributors may be used\n": "BSD-3-Clause",
		"nothing recognisable here\n": "",
		"":                            "",
	} {
		if got := detectSpdx(text); got != want {
			t.Errorf("detectSpdx(%.48q) = %q, want %q", text, got, want)
		}
	}
}

// The ladder is ordered, and more than one rule only gives the right answer because of it.
func TestDetectSpdxLadderOrder(t *testing.T) {
	// 0BSD's own text opens with ISC's distinctive sentence; 0BSD is tested first.
	zero := "Zero-Clause BSD\n\nPermission to use, copy, modify, and/or distribute this software " +
		"for any purpose with or without fee is hereby granted.\n"
	if got := detectSpdx(zero); got != "0BSD" {
		t.Errorf("0BSD text detected as %q", got)
	}
	// The affero and lesser spellings both contain the plainer names; their rules come first.
	if got := detectSpdx("GNU Affero General Public License\nVersion 3\n"); got != "AGPL-3.0-only" {
		t.Errorf("AGPL text detected as %q", got)
	}
	if got := detectSpdx("GNU Lesser General Public License\nVersion 3\n"); got != "LGPL-3.0-only" {
		t.Errorf("LGPL text detected as %q", got)
	}
	// A BSD body naming its contributors clause is 3-clause; the same body without it is 2.
	body := "Redistribution and use in source and binary forms are permitted.\n"
	if got := detectSpdx(body); got != "BSD-2-Clause" {
		t.Errorf("plain BSD body detected as %q", got)
	}
	if got := detectSpdx(body + "The names of its contributors may be used to endorse products.\n"); got != "BSD-3-Clause" {
		t.Errorf("BSD body with a contributors clause detected as %q", got)
	}
}

// Only the first 40 LINES are examined — a license appended below a long preamble is not
// detected, and that boundary has to fall in the same place as the TypeScript's slice(0, 40).
func TestDetectSpdxReadsOnlyTheHead(t *testing.T) {
	filler := strings.Repeat("x\n", 39) // lines 0..38
	if got := detectSpdx(filler + "MIT License\n"); got != "MIT" {
		t.Errorf("keyword on line 40 = %q, want MIT", got)
	}
	if got := detectSpdx(filler + "x\nMIT License\n"); got != "" {
		t.Errorf("keyword on line 41 = %q, want no detection", got)
	}
}

// The head is flattened with JavaScript's notion of whitespace, which is not Go's. A license
// pasted out of a word processor separates its words with U+00A0, and JavaScript also counts
// U+FEFF — get the set wrong and the Go build alone stops recognising the file.
func TestDetectSpdxCollapsesJavaScriptWhitespace(t *testing.T) {
	for _, sep := range []string{
		" ", "   ", "\t\t", "\r", "\f", "\v",
		"\u00a0", // NBSP: JavaScript whitespace, not Go's
		"\u2009", // thin space
		"\u3000", // ideographic space
		"\ufeff", // BOM: JavaScript whitespace, not Go's (a literal one is illegal here)
	} {
		if got := detectSpdx("MIT" + sep + "License\n"); got != "MIT" {
			t.Errorf("separator %q: detectSpdx = %q, want MIT", sep, got)
		}
	}
	// U+0085 NEL is the other direction: Go's strings.TrimSpace treats it as space and
	// JavaScript does not, so it must NOT collapse — the two words stay glued together.
	if got := detectSpdx("MIT\u0085License\n"); got != "" {
		t.Errorf("U+0085 collapsed like whitespace: detectSpdx = %q, want no detection", got)
	}
}

// CRLF is one line separator, not two: a Windows checkout must detect the same id, and its
// lines have to be counted the same way for the 40-line cap.
func TestDetectSpdxHandlesCRLF(t *testing.T) {
	if got := detectSpdx("MIT License\r\n\r\nCopyright (c) 2024\r\n"); got != "MIT" {
		t.Errorf("CRLF license = %q, want MIT", got)
	}
	if got := detectSpdx(strings.Repeat("x\r\n", 40) + "MIT License\r\n"); got != "" {
		t.Errorf("CRLF keyword past line 40 = %q, want no detection", got)
	}
	// A lone CR is not a line break, but it IS whitespace, so it collapses to a space.
	if got := detectSpdx("MIT\rLicense\n"); got != "MIT" {
		t.Errorf("lone CR = %q, want MIT", got)
	}
}

func TestResolveLicense(t *testing.T) {
	if resolveLicense(nil, "") != nil {
		t.Error("no file and no override should be no license")
	}
	fromConfig := resolveLicense(nil, "MIT")
	if fromConfig.Source != "config" || fromConfig.File != nil || *fromConfig.Spdx != "MIT" {
		t.Errorf("config license = %+v", fromConfig)
	}
	// A configured id wins, but the detected file is still recorded so the page can link it.
	both := resolveLicense(&detectedLicense{File: "LICENSE", Spdx: "Apache-2.0"}, "MIT")
	if *both.Spdx != "MIT" || both.File == nil || *both.File != "LICENSE" || both.Source != "config" {
		t.Errorf("layered license = %+v", both)
	}
	// A file we could not identify is still a license: the row links the text, the id is null.
	unknown := resolveLicense(&detectedLicense{File: "COPYING", Spdx: ""}, "")
	if unknown.Spdx != nil || unknown.Source != "file" || *unknown.File != "COPYING" {
		t.Errorf("unrecognised license text = %+v", unknown)
	}
	detected := resolveLicense(&detectedLicense{File: "LICENSE.md", Spdx: "MIT"}, "")
	if *detected.Spdx != "MIT" || *detected.File != "LICENSE.md" || detected.Source != "file" {
		t.Errorf("detected license = %+v", detected)
	}
}
